File: python/src/contriwork_notifications/adapters/telegram.py
"""Telegram Bot API adapter — HTTPS POST to api.telegram.org/bot<token>/sendMessage."""

from __future__ import annotations

from typing import Any

import httpx

from ..adapter import AdapterDeliverResult, AdapterStatus
from ..errors import ErrorCode
from ..payload import Payload
from ..severity import Severity

_DEFAULT_API_BASE = "https://api.telegram.org"
_DEFAULT_TIMEOUT_S = 10.0

_GENERIC_FAILURE = "Notification delivery failed"
_GENERIC_AUTH = "Authentication rejected"


class TelegramAdapter:
    """Telegram Bot API delivery via HTTPS POST.

    The bot token is opaque to the package (consumer's responsibility to load
    from its own config source). Messages are sent as plain text to a single
    pre-configured chat id; multi-chat fan-out is the consumer's job.

    Silent mode (active quiet hours per the client config): forwarded as the
    Telegram Bot API ``disable_notification: true`` flag.
    """

    def __init__(
        self,
        bot_token: str,
        chat_id: str | int,
        *,
        name: str = "telegram",
        api_base: str = _DEFAULT_API_BASE,
        timeout_seconds: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        self._bot_token = bot_token
        self._chat_id = chat_id
        self.name = name
        self._api_base = api_base
        self._timeout = timeout_seconds

    async def is_available(self) -> bool:
        return bool(self._bot_token) and self._chat_id not in (None, "", 0)

    async def deliver(
        self,
        severity: Severity,
        payload: Payload,
        silent: bool,
    ) -> AdapterDeliverResult:
        del severity  # plain-text delivery; severity is not encoded in the wire payload
        url = f"{self._api_base}/bot{self._bot_token}/sendMessage"
        text = f"{payload.title}\n\n{payload.body}"
        if payload.url is not None:
            label = payload.url_title or payload.url
            text += f"\n\n{label}: {payload.url}"
        body: dict[str, Any] = {
            "chat_id": self._chat_id,
            "text": text,
            "disable_notification": silent,
        }

        try:
            async with httpx.AsyncClient(timeout=self._timeout) as client:
                response = await client.post(url, json=body)
        except httpx.TimeoutException:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.TIMEOUT,
                detail=_GENERIC_FAILURE,
            )
        except httpx.HTTPError:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.UPSTREAM_ERROR,
                detail=_GENERIC_FAILURE,
            )

        return _map_response(response.status_code)


def _map_response(status_code: int) -> AdapterDeliverResult:
    if 200 <= status_code < 300:
        return AdapterDeliverResult(status=AdapterStatus.DELIVERED)
    if status_code == 429:
        return AdapterDeliverResult(
            status=AdapterStatus.RETRIABLE_FAILURE,
            error_code=ErrorCode.RATE_LIMITED,
            detail=_GENERIC_FAILURE,
        )
    if status_code in (401, 403):
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.AUTH_FAILED,
            detail=_GENERIC_AUTH,
        )
    if 400 <= status_code < 500:
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.ADAPTER_INVALID_PAYLOAD,
            detail=_GENERIC_FAILURE,
        )
    return AdapterDeliverResult(
        status=AdapterStatus.RETRIABLE_FAILURE,
        error_code=ErrorCode.UPSTREAM_ERROR,
        detail=_GENERIC_FAILURE,
    )

File: typescript/src/adapters/telegram.ts
import type { Adapter, AdapterDeliverResult } from "../adapter";
import { AdapterStatus } from "../adapter";
import { ErrorCode } from "../errors";
import type { Payload } from "../payload";
import type { Severity } from "../severity";

const DEFAULT_API_BASE = "https://api.telegram.org";
const DEFAULT_TIMEOUT_MS = 10_000;
const GENERIC_FAILURE = "Notification delivery failed";
const GENERIC_AUTH = "Authentication rejected";

export interface TelegramOptions {
  readonly botToken: string;
  readonly chatId: string;
  readonly name?: string;
  readonly apiBase?: string;
  readonly timeoutMs?: number;
  readonly fetchImpl?: typeof fetch;
}

/**
 * Telegram Bot API delivery via HTTPS POST to
 * `<apiBase>/bot<TOKEN>/sendMessage`. Plain-text body; severity is not
 * encoded in the wire payload. Silent mode forwards the Bot API
 * `disable_notification: true` flag.
 */
export class TelegramAdapter implements Adapter {
  readonly name: string;
  private readonly botToken: string;
  private readonly chatId: string;
  private readonly apiBase: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: TelegramOptions) {
    this.botToken = options.botToken;
    this.chatId = options.chatId;
    this.name = options.name ?? "telegram";
    this.apiBase = options.apiBase ?? DEFAULT_API_BASE;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
  }

  async isAvailable(): Promise<boolean> {
    return Boolean(this.botToken) && Boolean(this.chatId);
  }

  async deliver(
    _severity: Severity,
    payload: Payload,
    silent: boolean,
  ): Promise<AdapterDeliverResult> {
    const url = `${this.apiBase}/bot${this.botToken}/sendMessage`;
    let text = `${payload.title}\n\n${payload.body}`;
    if (payload.url !== undefined) {
      const label = payload.urlTitle ?? payload.url;
      text += `\n\n${label}: ${payload.url}`;
    }
    const body = JSON.stringify({
      chat_id: this.chatId,
      text,
      disable_notification: silent,
    });

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body,
        signal: controller.signal,
      });
    } catch {
      if (controller.signal.aborted) {
        return {
          status: AdapterStatus.RetriableFailure,
          errorCode: ErrorCode.Timeout,
          detail: GENERIC_FAILURE,
        };
      }
      return {
        status: AdapterStatus.RetriableFailure,
        errorCode: ErrorCode.UpstreamError,
        detail: GENERIC_FAILURE,
      };
    } finally {
      clearTimeout(timer);
    }

    return mapResponse(response.status);
  }
}

function mapResponse(statusCode: number): AdapterDeliverResult {
  if (statusCode >= 200 && statusCode < 300) {
    return { status: AdapterStatus.Delivered };
  }
  if (statusCode === 429) {
    return {
      status: AdapterStatus.RetriableFailure,
      errorCode: ErrorCode.RateLimited,
      detail: GENERIC_FAILURE,
    };
  }
  if (statusCode === 401 || statusCode === 403) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AuthFailed,
      detail: GENERIC_AUTH,
    };
  }
  if (statusCode >= 400 && statusCode < 500) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AdapterInvalidPayload,
      detail: GENERIC_FAILURE,
    };
  }
  return {
    status: AdapterStatus.RetriableFailure,
    errorCode: ErrorCode.UpstreamError,
    detail: GENERIC_FAILURE,
  };
}
